src/nunavut/lang/cpp/templates/_fields_as_union.j2

```diff
@@ -1,17 +1,22 @@
-    /*
-    TODO: finish me. intrinsic support for C++ is an incomplete feature. Watch
-          https://github.com/UAVCAN/nunavut/issues/91 for progress.
-    {{ composite_type.tag_field_type | declaration }} _tag_;
-    union {
+{#-
+ # Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ # Copyright (C) 2021  UAVCAN Development Team  <uavcan.org>
+ # This software is distributed under the terms of the MIT License.
+#}
+    union // The union is placed first to ensure that the active element address equals the object address.
```

Comment:
I am a little uneasy about this implementation because it violates AUTOSAR/MISRA (which prohibit unions) and is generally below the safety bar attainable with sensible C++ code. We spoke about it earlier and I would like to revisit this question now.

Beside the type safety issues, this specific implementation lacks the API handles necessary for constructing any non-trivial template code such as traversing/selecting the options at compile time without hard-coding the available options. This will become important, for instance, when implementing the Register interface, which relies on unions heavily.

We can't use `std::variant` due to our commitment to support old C++ standards, but maybe we can work around this with the help of an option switch that selects which `variant` type to use, similar to what you have for variable-length arrays? If the user wants C++17, then the generated code would simply use `std::variant`. Otherwise, the user would resort to a custom class or one of the several off-the-shelf implementations of the same, like https://github.com/martinmoene/variant-lite (there are others).

This should not increase the implementation workload on you since you will be able to rely on `std::variant` instead of reinventing an alternative type-safe mechanism.

Reply:
I vote for this for C++11 and 14 then switch over to std::variant when generating c++17. Would that be acceptable?

Reply:
no :) See, if you keep this, there will be two implementations with vastly different APIs. The current one will be mostly useless unless we equip it with compile-time introspection hooks provided by std::variant. I think it is easier to just use a third-party `variant` than reinvent this little wheel.

Reply:
poo. I really don't like the amount of Shim code we're introducing here. What 3p variant do you think would work?

Reply:
How about https://github.com/martinmoene/variant-lite (Boost license)

Reply:
Tracking in #91 TODO list. Let's get this big chunk of incremental work, which is behind an --experimental flag, out.